crates/store/src/state.rs

```diff
@@ -783,61 +751,110 @@ impl State {
+                let ValuePath {
+                    value: latest_state_commitment,
+                    path: proof,
+                } = inner.account_tree.open(&account_id.into());
+                (account_id, AccountWitness::new(latest_state_commitment, proof))
             })
-            .collect::<Result<_, AccountError>>()?;
+            .collect::<BTreeMap<AccountId, AccountWitness>>();
 
-        let nullifiers: Vec<NullifierWitness> = nullifiers
+        // Fetch witnesses for all nullifiers. We don't check whether the nullifiers are spent or
+        // not as this is done as part of proposing the block.
+        let nullifier_witnesses: BTreeMap<Nullifier, NullifierWitness> = nullifiers
             .iter()
+            .copied()
             .map(|nullifier| {
-                let proof = inner.nullifier_tree.open(nullifier);
-
-                NullifierWitness { nullifier: *nullifier, proof }
+                let proof = inner.nullifier_tree.open(&nullifier);
+                (nullifier, NullifierWitness::new(proof))
             })
             .collect();
 
-        let found_unauthenticated_notes =
-            self.get_note_authentication_info(unauthenticated_notes).await?;
+        // Release the lock.
+        std::mem::drop(inner);
 
-        Ok(BlockInputs {
-            block_header: latest,
-            chain_peaks,
-            account_states,
-            nullifiers,
-            found_unauthenticated_notes,
-        })
+        // Fetch the block headers for all blocks in the partial MMR plus the latest one which will
+        // be used as the previous block header of the block being built.
+        let mut headers = self
+            .db
+            .select_block_headers(blocks.into_iter().chain(std::iter::once(latest_block_number)))
+            .await
+            .map_err(GetBlockInputsError::SelectBlockHeaderError)?;
+
+        // Find and remove the latest block as we must not add it to the chain MMR, since it is
+        // not yet in the chain.
+        let latest_block_header_index = headers
+            .iter()
+            .enumerate()
+            .find_map(|(index, header)| {
+                (header.block_num() == latest_block_number).then_some(index)
+            })
+            .expect("DB should have returned the header of the latest block header");
+
+        // The order doesn't matter for ChainMmr::new, so swap remove is fine.
+        let latest_block_header = headers.swap_remove(latest_block_header_index);
+
+        // SAFETY: This should not error because:
+        // - we're passing exactly the block headers that we've added to the partial MMR,
+        // - so none of the block header's block numbers should exceed the chain length of the
+        //   partial MMR,
+        // - and we've added blocks to a BTreeSet, so there can be no duplicates.
+        let chain_mmr = ChainMmr::new(partial_mmr, headers)
+            .expect("partial mmr and block headers should be consistent");
```

Comment:
If for some reason, the DB does not return all requested block headers this would panic. Is that consistent with error handling in the node or should we actually return an error here and for similar `expect`s and let higher level components deal with it?

Reply:
This would imply db corruption, or another writer somehow correct?

I think I'm happy with a panic here - given that we're essentially screwed if this occurs right?